feat: 이메일 인증, 아이디 찾기, 비밀번호 초기화 구현, 로그인 레디스 토큰 사용으로 변경

zhy2on opened the pull request:
## #️⃣ 관련 이슈
close #10 #8 #14 

## 📝작업 내용
  - [POST] `/api/email-auth/send-code`: 이메일로 인증 코드를 전송. 재전송 5분 제한.
  - [POST] `/api/email-auth/verify-code`: 사용자가 입력한 인증 코드를 확인.
  - [POST] `api/auth/find-username`:  이메일에 해당하는 사용자 아이디 찾기
  - [POST[ `/api/auth//reset-password`: 아이디, 이메일 입력을 받고 사용자의 비밀번호를 랜덤값으로 초기화 한 후 이메일로 전송
  - [DELETE] `api/auth/user`: 회원탈퇴. 로그인한 상태에서만 가능하도록 access token 헤더가 필요.

## 💬향후 개선사항

이메일 템플릿 개선 필요

zhy2on committed:
feat: mail, redis 설정 추가

zhy2on committed:
feat: mail, redis 관련 유틸 추가

zhy2on committed:
feat: 이메일 인증 프로세스 구현
- redis에 이메일 인증 관련 키 값 저장
- 이메일 인증이 완료된 사용자만 회원가입이 되게 변경

zhy2on committed:
feat: send-code api에 이미 존재하는 메일 검증 추가

zhy2on committed:
feat: email auth api 관련 Swagger 설정 추가

zhy2on committed:
feat: 이메일 인증 프로세스 추가에 맞춰 auth 테스트 수정

zhy2on committed:
feat: email auth 테스트 코드 추가

zhy2on committed:
chore: domain->dto 로 패키지명 변경

zhy2on committed:
chore: 불필요한 코드 제거

zhy2on committed:
refactor: refresh token을 redis에 저장하도록 변경 #14

zhy2on committed:
feat: 아이디 찾기 구현 #8
- email로 아이디 찾기 구현
- userRepository에 validation에 필요한 method 추가 구현

zhy2on committed:
refactor: 아이디 찾기 get 요청으로 변경 #8

zhy2on committed:
feat: 비밀번호 초기화 구현 #8
- 랜덤한 값으로 비밀번호 설정 후 사용자 메일로 전송

BOK-MINJEONG commented:
인증 코드 재전송할 때 시간 제한을 걸면 좋을 것 같습니다. 예를 들어, 5분에 한 번씩만 메일을 전송할 수 있도록 방어 로직을 추가하는거도 괜찮은 거 같아요!

BOK-MINJEONG reviewed (commented):
인증 코드 재전송할 때 시간 제한을 걸면 좋을 것 같습니다. 예를 들어, 5분에 한 번씩만 메일을 전송할 수 있도록 방어 로직을 추가하는거도 괜찮은 거 같아요!

zhy2on commented:
> 인증 코드 재전송할 때 시간 제한을 걸면 좋을 것 같습니다. 예를 들어, 5분에 한 번씩만 메일을 전송할 수 있도록 방어 로직을 추가하는거도 괜찮은 거 같아요!

추가해서 올리도록 하겠습니다!

zhy2on committed:
feat: 회원 탈퇴 api 구현 및 로그아웃, 토큰 갱신 api 리팩토링
- 회원탈퇴, 로그아웃, 토큰 갱신 모두 인증이 필요
- spring security에서 인증된 내용을 UserDetails 객체로 사용할 수 있음
- AuthenticationPrincipal UserDetails를 사용하여 리팩토링
- SecurityConfig를 수정하여 인증이 필요한 api를 정확히 구분

zhy2on committed:
fix: dto로 잘못 변경했던 것 entity로 변경

zhy2on committed:
feat: request에 @NotBlank로 유효성 검증 추가

zhy2on committed:
refactor: 불필요한 추가 검증 뺌
- NotNull을 선언했기 때문에 nullable=false가 불필요해 보여 제거

zhy2on committed:
refactor: 불필요한 Transactional 어노테이션 제거

zhy2on committed:
refactor: request dto에서 입력값 검증
- `@Size`, `@Email`등을 활용하여 입력값 검증 메타데이터 설정
- 필요없어진 EmailValidator 제거
- `@Valid`를 추가하여 입력값 검증이 이뤄지게 함

zhy2on committed:
feat: 이메일 재전송 시간 제한 기능 추가

zhy2on committed:
feat: 이메일 재전송 제한 시간 추가

zhy2on committed:
feat: login response에 username 포함

zhy2on committed:
Resolve merge conflict

BOK-MINJEONG reviewed (approved):
구현할 기능 많으셨을 텐데 수고 많으셨습니다ㅠ
이메일 인증 기능은 한 번도 해보지 않았는데 코드 리뷰하면서 많이 배웠습니다! 감사합니다 :)
